### PR TITLE
Fix EVR regex to not match email addresses with - in them

### PR DESCRIPTION
**Pull request by dcermak**
TODO:

- [x] Write new tests or update the old ones to cover new functionality.
- [x] Update doc-strings where appropriate.
- [x] Update or write new documentation in `packit/packit.dev`.

### CODE REVIEW

File: specfile/changelog.py [hunk 1]
```diff
@@ -88,12 +88,12 @@ def evr(self) -> Optional[str]:
...
             ^.*
-            \s+                       # preceding whitespace
-            ((?P<sb>\[)|(?P<rb>\())?  # optional opening bracket
-            (?P<evr>(\d+:)?\S+-\S+?)  # EVR
-            (?(sb)\]|(?(rb)\)))       # matching closing bracket
-            :?                        # optional colon
-            \s*                       # optional following whitespace
+            \s+                             # preceding whitespace
+            ((?P<sb>\[)|(?P<rb>\())?        # optional opening bracket
+            (?P<evr>(\d+:)?(\d+\.)\S*-\S+?) # EVR
```
[the remark this answers is not in the document]
Comment (nforro) [reply specfile/changelog.py:93]: > I think it would be better to ditch the regex here and instead use the [EVR class](https://github.com/packit/specfile/blob/c0a98a85d9e43f59447248bb7f5258c2406ce9f1/specfile/utils.py#L17) that has been implemented meanwhile.

Actually that won't solve the problem as an e-mail address can be a valid EVR string :sweat_smile:

[the remark this answers is not in the document]
Comment (nforro) [reply specfile/changelog.py:93]: > Any suggestions what to do here? The added test case is pretty obviously not containing an EVR and imho we should try to return `None` in that case too.

I agree. AFAICT in Fedora spec files the EVR is either prefixed with ` - ` or enclosed in `[]` or `()`, and there is always an author (with our without an e-mail) preceding it. So it should be possible to differentiate between an entry with an author followed by an EVR and an entry with an author only.

> The bigger problem is, that openSUSE style detached changelogs **never** contain an EVR but the regex matches the email of the author and thereby it always returns something (see #444)

Why is that a problem? AFAICS the `evr` property is accessed only in the `filter()` method.

[the remark this answers is not in the document]
Comment (dcermak) [reply specfile/changelog.py:93]: > > Any suggestions what to do here? The added test case is pretty obviously not containing an EVR and imho we should try to return `None` in that case too.
> 
> I agree. AFAICT in Fedora spec files the EVR is either prefixed with `-` or enclosed in `[]` or `()`, and there is always an author (with or without an e-mail) preceding it. So it should be possible to differentiate between an entry with an author followed by an EVR and an entry with an author only.

Unfortunately that's not the case, as there's an example in the tests:
https://github.com/dcermak/specfile/blob/55039df973d5186f509b8047acf9e6c0e3b1ee6e/tests/unit/test_changelog.py#L19

> 
> > The bigger problem is, that openSUSE style detached changelogs **never** contain an EVR but the regex matches the email of the author and thereby it always returns something (see #444)
> 
> Why is that a problem? AFAICS the `evr` property is accessed only in the `filter()` method.

So far I have been unable to build a regex that works universally and I am hesitant to push broken code. I don't know where this is used (or will be used in the future).

[the remark this answers is not in the document]
Comment (nforro) [reply specfile/changelog.py:93]: > Unfortunately that's not the case, as there's an example in the tests:
> https://github.com/dcermak/specfile/blob/55039df973d5186f509b8047acf9e6c0e3b1ee6e/tests/unit/test_changelog.py#L19

Oh, right, I overlooked that one :slightly_frowning_face:

[the remark this answers is not in the document]
Comment (nforro) [reply specfile/changelog.py:93]: For example this isn't perfect, but seems to work for all test cases:
```python
    @property
    def evr(self) -> Optional[str]:
        """EVR (epoch, version, release) of the entry."""
        m = re.match(
            r"""
            ^.*
            \d{4}                     # year
            \s+                       # whitespace
            .+                        # author
            \s+                       # preceding whitespace
            ((?P<sb>\[)|(?P<rb>\())?  # optional opening bracket
            (?P<evr>\S+?)             # EVR
            (?(sb)\]|(?(rb)\)))       # matching closing bracket
            :?                        # optional colon
            \s*                       # optional following whitespace
            $
            """,
            self.header,
            re.VERBOSE,
        )
        if not m:
            return None
        evr = m.group("evr")
        try:
            EVR.from_string(evr)
        except SpecfileException:
            return None
        if evr.startswith("<") and evr.endswith(">"):
            # it's most likely an e-mail
            return None
        return evr
```

[the remark this answers is not in the document]
Comment (dcermak) [reply specfile/changelog.py]: Done, thanks for the suggestion!

File: specfile/changelog.py [hunk 2]
```diff
@@ -287,22 +305,20 @@ def parse_evr(s):
         if since is None:
             start_index = 0
         else:
+            since_evr: EVR = parse_evr(since) if isinstance(since, str) else since
```
Comment (nforro) [inline specfile/changelog.py:308]: I think it would be nicer to extend `parse_evr()` to accept `EVR` or string and avoid this.

File: tests/unit/test_changelog.py [hunk 3]
```diff
@@ -39,8 +45,12 @@
...
     ],
 )
-def test_entry_evr(header, evr):
-    assert ChangelogEntry(header, [""]).evr == evr
+def test_entry_evr(header, evr_str):
+    evr = ChangelogEntry(header, [""]).evr
+    if evr_str:
+        assert evr == EVR.from_string(evr_str)
+    else:
+        assert evr is None
```
Comment (nforro) [inline tests/unit/test_changelog.py:53]: Wouldn't it make more sense to parametrize the test with `EVR` objects (or `None`), like e.g. here?

https://github.com/packit/specfile/blob/c0a98a85d9e43f59447248bb7f5258c2406ce9f1/tests/unit/test_utils.py#L91-L100

It's more verbose and a bit more work, but I like it more that way :slightly_smiling_face:

File: specfile/changelog.py [hunk 4]
```diff
@@ -83,25 +100,24 @@ def __repr__(self) -> str:
...
+        evr_s = m.group("evr")
+        try:
+            evr = EVR.from_string(evr_s)
+        except SpecfileException:
+            return None
+
+        # looks like an email
+        if evr_s.startswith("<") and evr_s.endswith(">"):
+            return None
+
```
Comment (nforro) [inline specfile/changelog.py:119]: I was thinking, if we get here, we could also check if the string contains a `@` or a number, EVR is not likely to contain a `@`  and on the other hand is likely to contain a number, unlike a surname for example. But maybe that's too much guessing.

Comment (dcermak) [reply specfile/changelog.py:119]: I wouldn't check for an `@`, some people obscure their emails like this one: `mszpak ATT wp DOTT pl`

**Review by nforro — APPROVED** [review]
Thanks!
